File: src/main/java/com/imralav/gmtools/battletracker/views/BattleTrackerRowView.java
package com.imralav.gmtools.battletracker.views;

import com.imralav.gmtools.battletracker.model.BattleTrackerRow;
import com.imralav.gmtools.battletracker.model.BattleTrackerUnit;
import com.imralav.gmtools.utils.ViewsLoader;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.control.Label;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.FlowPane;
import javafx.scene.layout.HBox;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

import java.io.IOException;
import java.util.Objects;
import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

@Slf4j
public class BattleTrackerRowView extends HBox {
    private static final String VIEW_PATH = "battletracker/units/row.fxml";

    @Getter
    private final BattleTrackerRow battleTrackerRow;

    @FXML
    private Label initiative;

    @FXML
    private FlowPane units;

    private Optional<BiConsumer<BattleTrackerUnit, BattleTrackerUnitView>> onUnitClickedAction;

    @Setter
    private Consumer<BattleTrackerRow> removeRowAction;

    BattleTrackerRowView(BattleTrackerRow battleTrackerRow) throws IOException {
        this(battleTrackerRow, null);
    }

    BattleTrackerRowView(BattleTrackerRow battleTrackerRow, BiConsumer<BattleTrackerUnit, BattleTrackerUnitView> onUnitClickedAction) throws IOException {
        FXMLLoader fxmlLoader = ViewsLoader.getViewLoader(VIEW_PATH);
        fxmlLoader.setRoot(this);
        fxmlLoader.setController(this);
        fxmlLoader.load();
        this.battleTrackerRow = battleTrackerRow;
        initiative.textProperty().bind(battleTrackerRow.initiativeProperty().asString());
        setOnUnitClickedAction(onUnitClickedAction);
        battleTrackerRow.getUnits().forEach(unit -> {
            try {
                BattleTrackerUnitView unitView = new BattleTrackerUnitView(unit);
                unitView.addEventHandler(MouseEvent.MOUSE_CLICKED, event -> {
                    this.onUnitClickedAction.ifPresent(action -> action.accept(unit, unitView));
                });
                unitView.setRemoveUnitAction(removedUnit -> {
                    boolean noUnitsInRow = this.battleTrackerRow.removeUnitAndCheckIfEmpty(removedUnit);
                    if (noUnitsInRow) {
                        removeRow();
                    }
                });
                units.getChildren().add(unitView);
            } catch (IOException e) {
                log.error("Couldn't create new unit view: {}", e.getMessage(), e);
            }
        });
        battleTrackerRow.selectedProperty().addListener((observable, oldValue, newValue) -> {
            if (newValue) {
                this.getStyleClass().add("selected");
            } else {
                this.getStyleClass().remove("selected");
            }
        });
        battleTrackerRow.registerRemovedUnitsListener(removedUnits -> {
            removedUnits.forEach(removedUnit -> {
                units.getChildren().removeIf(unitView -> {
                    return removedUnit == ((BattleTrackerUnitView) unitView).getBattleTrackerUnit();
                });
            });
        });
    }

    public void setOnUnitClickedAction(BiConsumer<BattleTrackerUnit, BattleTrackerUnitView> onUnitClickedAction) {
        this.onUnitClickedAction = Optional.ofNullable(onUnitClickedAction);
    }

    @FXML
    public void removeRow() {
        if (Objects.nonNull(removeRowAction)) {
            removeRowAction.accept(battleTrackerRow);
        }
    }
}

File: gui/src/main/kotlin/com/imralav/gmtools/gui/battletracker/views/BattleTrackerRowView.kt
package com.imralav.gmtools.gui.battletracker.views

import com.imralav.gmtools.gui.battletracker.model.BattleTrackerRow
import com.imralav.gmtools.gui.battletracker.model.BattleTrackerUnit
import com.imralav.gmtools.gui.utils.ViewsLoader
import javafx.beans.value.ObservableValue
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.Node
import javafx.scene.control.Label
import javafx.scene.input.MouseEvent
import javafx.scene.layout.FlowPane
import javafx.scene.layout.HBox
import java.net.URL
import java.util.Optional
import java.util.ResourceBundle
import java.util.function.BiConsumer
import java.util.function.Consumer

class BattleTrackerRowView @JvmOverloads internal constructor(val battleTrackerRow: BattleTrackerRow, onUnitClickedAction: BiConsumer<BattleTrackerUnit?, BattleTrackerUnitView?>? = null) : HBox(), Initializable {

    @FXML
    private lateinit var initiative: Label

    @FXML
    private lateinit var units: FlowPane

    private var onUnitClickedAction: Optional<BiConsumer<BattleTrackerUnit?, BattleTrackerUnitView?>>? = Optional.ofNullable(onUnitClickedAction)

    var removeRowAction: Consumer<BattleTrackerRow>? = null

    fun setOnUnitClickedAction(onUnitClickedAction: BiConsumer<BattleTrackerUnit?, BattleTrackerUnitView?>?) {
        this.onUnitClickedAction = Optional.ofNullable(onUnitClickedAction)
    }

    @FXML
    fun removeRow() = removeRowAction?.accept(battleTrackerRow)

    init {
        val fxmlLoader = ViewsLoader.getViewLoader("battletracker/units/row.fxml")
        fxmlLoader.setRoot(this)
        fxmlLoader.setController(this)
        fxmlLoader.load<Any>()
    }

    override fun initialize(location: URL?, resources: ResourceBundle?) {
        initiative.textProperty().bind(battleTrackerRow.initiativeProperty().asString())
        battleTrackerRow.getUnits().forEach(Consumer { unit: BattleTrackerUnit? ->
            val unitView = BattleTrackerUnitView(unit!!)
            unitView.addEventHandler(MouseEvent.MOUSE_CLICKED) { event: MouseEvent? -> this.onUnitClickedAction!!.ifPresent { action: BiConsumer<BattleTrackerUnit?, BattleTrackerUnitView?> -> action.accept(unit, unitView) } }
            unitView.removeUnitAction = Consumer { removedUnit: BattleTrackerUnit? ->
                val noUnitsInRow = this.battleTrackerRow.removeUnitAndCheckIfEmpty(removedUnit)
                if (noUnitsInRow) {
                    removeRow()
                }
            }
            units.children.add(unitView)
        })
        battleTrackerRow.selectedProperty().addListener { observable: ObservableValue<out Boolean>?, oldValue: Boolean?, newValue: Boolean ->
            if (java.lang.Boolean.TRUE == newValue) {
                styleClass.add("selected")
            } else {
                styleClass.remove("selected")
            }
        }
        battleTrackerRow.registerRemovedUnitsListener(Consumer { removedUnits: List<BattleTrackerUnit> -> removedUnits.forEach(Consumer { removedUnit: BattleTrackerUnit -> units.children.removeIf { unitView: Node -> removedUnit == (unitView as BattleTrackerUnitView).battleTrackerUnit } }) })
    }
}